model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remotestore/RemoteCacheSessionsLoader.java

```diff
@@ -79,23 +81,36 @@ public WorkerResult loadSessions(KeycloakSession session, RemoteCacheSessionsLoa
         RemoteCache<?, ?> remoteCache = getRemoteCache(session);
 
         int countLoaded = 0;
-        try (CloseableIterator<Map.Entry<Object, Object>> it = remoteCache.retrieveEntries(null, loaderContext.getSessionsPerSegment())) {
+        try (CloseableIterator<Map.Entry<Object, MetadataValue<Object>>> it = remoteCache.retrieveEntriesWithMetadata(null, loaderContext.getSessionsPerSegment())) {
             Map<Object, Object> toInsert = new HashMap<>(loaderContext.getSessionsPerSegment());
             int count = 0;
+            int maxLifespan = -1;
+            int maxIdle = -1;
             while (it.hasNext()) {
-                Map.Entry<?,?> entry = it.next();
-                toInsert.put(entry.getKey(), entry.getValue());
+                Map.Entry<Object, MetadataValue<Object>> entry = it.next();
+                if (entry.getValue().getLifespan() > 0) {
+                    // Calculate the remaining lifetime reduced by the current time, not Keycloak time as the remote Infinispan isn't on Keycloak's clock.
+                    // The lifetime will be larger than on the remote store for those entries, but all sessions contain timestamp which will be validated anyway.
+                    // If we don't trust the clock calculations here, we would instead use the maxLifeSpan as is, which could enlarge the expiry time significantly.
+                    maxLifespan = Math.max(maxLifespan, Math.max(0, entry.getValue().getLifespan() - (int) ((System.currentTimeMillis() - entry.getValue().getCreated()) / 1000)));
+                }
+                if (entry.getValue().getMaxIdle() > 0) {
+                    // The max idle time on the remote store is set to the max lifetime as remote store entries are not touched on read, and therefore would otherwise expire too early.
+                    // Still, this is the only number we have available, so we use it.
+                    maxIdle = Math.max(maxIdle, entry.getValue().getMaxIdle());
```

Comment:
When we could do those assumptions about the remote store, then this could be optimized. Still I would like to avoid additional assumptions what other pieces of the code do. 

I don't think we can do any better here, so I suggest to keep it as is for now.